src/vcpkg/base/system.mac.cpp

```diff
@@ -0,0 +1,251 @@
+            // AF_PACKET and sockaddr_ll: https://man7.org/linux/man-pages/man7/packet.7.html
+            // AF_LINK and sockaddr_dl: https://illumos.org/man/3SOCKET/sockaddr_dl
+#if defined(__linux__)
+            auto address = reinterpret_cast<sockaddr_ll*>(interface->ifa_addr);
+            if (address->sll_halen != MAC_BYTES_LENGTH) continue;
+            std::memcpy(bytes, address->sll_addr, MAC_BYTES_LENGTH);
+#elif defined(__APPLE__)
+            auto address = reinterpret_cast<sockaddr_dl*>(interface->ifa_addr);
+            if (address->sdl_alen != MAC_BYTES_LENGTH) continue;
+            // The macro LLADDR() returns the start of the link-layer network address.
+            std::memcpy(bytes, LLADDR(address), MAC_BYTES_LENGTH);
+#endif
+            auto maybe_mac = mac_bytes_to_string(Span<unsigned char>(bytes, MAC_BYTES_LENGTH));
+            if (is_valid_mac_for_telemetry(maybe_mac))
+            {
+                return Hash::get_string_hash(maybe_mac, Hash::Algorithm::Sha256);
+            }
+        }
+#else
+        // fallback when getifaddrs() is not available
+        struct socket_guard
+        {
+            int fd = -1;
+            ~socket_guard()
+            {
+                if (fd > 0) close(fd);
+            }
+            inline operator int() { return fd; }
+        } fd{socket(AF_INET, SOCK_DGRAM, IPPROTO_IP)};
+
+        if (fd == -1) return "0";
+
+        ifconf interfaces;
+        std::memset(&interfaces, 0, sizeof(ifconf));
+        // Calling ioctil with ifconf.ifc_req set to null returns
+        // the size of buffer needed to contain all interfaces
+        // in ifconf.ifc_len.
+        // https://www.man7.org/linux/man-pages/man7/netdevice.7.html
+        if (ioctl(fd, SIOCGIFCONF, &interfaces) < 0) return "0";
+
+        auto data = std::vector<char>(interfaces.ifc_len);
+        interfaces.ifc_len = data.size();
+        interfaces.ifc_buf = data.data();
+        if (ioctl(fd, SIOCGIFCONF, &interfaces) < 0) return "0";
+
+        // On a successful call, ifc_req contains a pointer to an array
+        // of ifreq structures filled with all currently active interface addresses.
+        // Within each structure ifr_name will receive the interface name, and
+        // ifr_addr the addres.
+        const ifreq* const end = interfaces.ifc_req + (interfaces.ifc_len / sizeof(ifreq));
+        unsigned char bytes[MAC_BYTES_LENGTH];
+        for (auto it = interfaces.ifc_req; it != end; ++it)
+        {
+            ifreq& interface = *it;
+
+            // Retrieve interface hardware addresses (ignore loopback)
+            if ((ioctl(fd, SIOCGIFFLAGS, &interface) >= 0) && !(interface.ifr_flags & IFF_LOOPBACK) &&
+                (ioctl(fd, SIOCGIFHWADDR, &interface) >= 0))
+            {
+                std::memcpy(bytes, interface.ifr_hwaddr.sa_data, MAC_BYTES_LENGTH);
+                auto maybe_mac = mac_bytes_to_string(Span<unsigned char>(bytes, MAC_BYTES_LENGTH));
```

Comment:
```suggestion
                auto maybe_mac = mac_bytes_to_string(Span<unsigned char>(reinterpret_cast<unsigned char*>(interface.ifr_hwaddr.sa_data), MAC_BYTES_LENGTH));
```

Reply:
I'm not sure why you're adding a reinterpret_cast here: it compiled before with no cast which means they type is already correct, right?

Reply:
Oh, I see, you're getting the member rather than the raw bytes, OK.

Reply:
Maybe `mac_bytes_to_string` should take `Span<char` so that no cast is needed?

Reply:
done